CollisionOperator/CollisionProcessors/neutronCEimp_class.f90

```diff
@@ -452,43 +472,73 @@ subroutine cutoffs(self, p, collDat, thisCycle, nextCycle)
     type(collisionData), intent(inout)   :: collDat
     class(particleDungeon),intent(inout) :: thisCycle
     class(particleDungeon),intent(inout) :: nextCycle
+    real(defReal), dimension(3)          :: val
+    real(defReal)                        :: minWgt, maxWgt, avWgt
 
-    if (p % E < self % minE ) then
+    if (p % E < self % minE) then
       p % isDead = .true.
+    ! Weight Windows treatment
+    elseif (self % weightWindows) then
+      val = self % weightWindowsMap % at(p)
+      minWgt = val(1)
+      maxWgt = val(2)
+      avWgt  = val(3)
+
+      if ((p % w > maxWgt) .and. (maxWgt /= ZERO)) then
+        call self % split(p, thisCycle, maxWgt)
+      elseif (p % w < minWgt) then
+        call self % russianRoulette(p, avWgt)
+      end if
+
+    ! Splitting with fixed threshold
     elseif ((self % splitting) .and. (p % w > self % maxWgt)) then
-      call self % split(p, thisCycle)
+      call self % split(p, thisCycle, self % maxWgt)
+    ! Roulette with fixed threshold and survival weight
     elseif ((self % roulette) .and. (p % w < self % minWgt)) then
-      call self % russianRoulette(p)
-    endif
+      call self % russianRoulette(p, self % avWgt)
+    end if
 
   end subroutine cutoffs
 
   !!
   !! Perform Russian roulette on a particle
   !!
-  subroutine russianRoulette(self, p)
+  subroutine russianRoulette(self, p, avWgt)
     class(neutronCEimp), intent(inout) :: self
     class(particle), intent(inout)     :: p
+    real(defReal), intent(in)          :: avWgt
 
-    if (p % pRNG % get() < (ONE - p % w/self % avWgt)) then
+    if (p % pRNG % get() < (ONE - p % w/avWgt)) then
       p % isDead = .true.
     else
-      p % w = self % avWgt
+      p % w = avWgt
     end if
 
   end subroutine russianRoulette
 
   !!
   !! Split particle which has too large a weight
   !!
-  subroutine split(self, p, thisCycle)
+  subroutine split(self, p, thisCycle, maxWgt)
     class(neutronCEimp), intent(inout)    :: self
     class(particle), intent(inout)        :: p
     class(particleDungeon), intent(inout) :: thisCycle
-    integer(shortInt)                     :: mult,i
+    real(defReal), intent(in)             :: maxWgt
+    integer(shortInt)                     :: mult, n, i
+    real(defReal)                         :: prob
 
     ! This value must be at least 2
-    mult = ceiling(p % w/self % maxWgt)
+    n = floor(p % w/maxWgt)
+    prob = p % w/maxWgt - n
+
+    if (p % pRNG % get() < prob) then
+      mult = n + 1
+    else
+      mult = n
+    end if
```

Comment:
What about something along the lines of 
```Fortran
n = floor((p % w / maxWgt) + p % pRNG & get())
```
Which is how the probabilistic rounding is usually handled in case of fission site generation.